components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java

```diff
@@ -0,0 +1,209 @@
+ */
+public class APIErrorResponseHandler {
+
+    private static final Log log = LogFactory.getLog(APIErrorResponseHandler.class);
+
+    private static final String AUTH_HEADER_NAME = "WWW-Authenticate";
+    private static final String CORRELATION_ID_MDC = "Correlation-ID";
+    private static final String BAD_REQUEST_ERROR_MSG = "Your client has issued a malformed or illegal request.";
+    private static final String UNAUTHORIZED_ERROR_MSG = "Authorization failure. Authorization information was" +
+            " invalid or missing from your request.";
+    private static final String FORBIDDEN_ERROR_MSG = "Operation is not permitted. You do not have permissions to " +
+            "make this request.";
+
+
+    /**
+     * Generate the error response according to the relevant API endpoint and the HTTP status.
+     */
+    public static void handleErrorResponse(AuthenticationContext authenticationContext, Response response, int error,
+                                            Exception e) throws IOException {
+
+        if (log.isDebugEnabled() && e != null) {
+            log.debug("Authentication Error ", e);
+        }
+        StringBuilder value = new StringBuilder(16);
+        value.append("realm user=\"");
+        if (authenticationContext != null && authenticationContext.getUser() != null) {
+            value.append(authenticationContext.getUser().getUserName());
+        }
+        value.append('\"');
+        response.setHeader(AUTH_HEADER_NAME, value.toString());
+        if (response.getRequest() != null) {
+            String uri = response.getRequest().getRequestURI();
+            uri = removeTenantDetailFromURI(uri);
+            if (isRequestFromScim2Api(uri)) { // handles only scim 2.0 API error responses.
+                handleScim2ApiErrorResponse(response, error, e);
+            } else if (isRequestFromDCREndpoint(uri)) {
+                handleDCRApiErrorResponse(response, error, e);
+            } else {
+                handleErrorResponseForCommonAPIs(response, error, e);
+            }
+        } else { // if request is null, sending a common error message
+            handleErrorResponseForCommonAPIs(response, error, e);
+        }
+    }
+
+    private static String removeTenantDetailFromURI(String uri) {
+        if (uri.startsWith("/t")) {
+            String[] uriSplit = uri.split("/");
+            StringBuilder builder = new StringBuilder();
+            if (uriSplit.length > 3) {
+                for (int i = 3; i < uriSplit.length; i++) {
+                    builder.append('/');
+                    builder.append(uriSplit[i]);
+                }
+                return builder.toString();
+            }
+        }
+        return uri;
+    }
+
+
+    private static boolean isRequestFromScim2Api(String uri) {
+
+        if (uri == null) {
+            return false;
+        }
+        return uri.startsWith("/scim2");
+    }
```

Comment:
Let's simplify this.
eg: `return StringUtils.isNotBlank(uri) && uri.startsWith("/scim2")`

Reply:
Fixed with https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/183